### PR TITLE
More restore optimizations

### PR DESCRIPTION
**Pull request by proteusvacuum**
@calellowitz made a few more changes

CC: @snopoke 

I ran (i.e. timings also include all reports, and fixtures)
```py
from corehq.apps.users.models import CommCareUser
from corehq.apps.ota.views import get_restore_response
user = CommCareUser.get_by_username('101624@enikshay.commcarehq.org')
%prun restore = get_restore_response('enikshay', user, version='2.0', overwrite_cache=True)
```

before (with user `101624@enikshay.commcarehq.org`)
```
279502322 function calls (275633510 primitive calls) in 890.520 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    92867  232.743    0.003  613.947    0.007 models.py:727(_reverse_index_map)
  1445693  228.319    0.000  228.319    0.000 {method 'items' of 'dict' objects}
199164653  154.866    0.000  154.866    0.000 {method 'values' of 'dict' objects}
    52121   49.428    0.001   52.716    0.001 {method 'execute' of 'psycopg2.extensions.cursor' objects}
7578292/7578032    7.361    0.000    8.796    0.000 {isinstance}
       63    5.142    0.082    5.142    0.082 {_socket.getaddrinfo}
  3422630    4.262    0.000    4.269    0.000 {hasattr}
  4679936    3.937    0.000    3.937    0.000 {method 'append' of 'list' objects}
    12704    3.918    0.000    4.862    0.000 clean_owners.py:111(_add_unchecked_indices_to_be_checked)
302639/12834    3.795    0.000    8.229    0.001 ElementTree.py:899(_serialize_xml)
   839875    3.770    0.000    9.470    0.000 base_properties.py:71(__get__)
    24830    3.539    0.000    3.539    0.000 {method 'remove' of 'list' objects}
   641645    3.075    0.000    3.581    0.000 base_properties.py:20(__init__)
580534/528267    2.828    0.000   14.294    0.000 compiler.py:348(compile)
1944677/1915950    2.799    0.000    2.902    0.000 {getattr}
   732962    2.787    0.000    8.456    0.000 expressions.py:667(get_db_converters)
  1031313    2.702    0.000    5.101    0.000 compiler.py:331(quote_name_unless_alias)
    54333    2.480    0.000    5.664    0.000 base.py:457(__init__)
2586906/2196750    2.318    0.000   70.212    0.000 {len}
    25715    2.303    0.000   21.225    0.001 compiler.py:165(get_select)
    12834    2.213    0.000    7.798    0.001 ElementTree.py:831(_namespaces)
    51343    2.147    0.000   11.267    0.000 compiler.py:764(get_converters)
    51141    2.105    0.000   50.316    0.001 query.py:1233(__iter__)
  1197943    2.104    0.000    2.106    0.000 {built-in method __new__ of type object at 0x919120}
    25707    2.095    0.000    6.395    0.000 compiler.py:475(get_default_columns)
400912/1913    2.076    0.000    5.470    0.003 copy.py:145(deepcopy)
   141727    2.072    0.000    8.413    0.000 inspect.py:895(getcallargs)
   759286    2.060    0.000    2.563    0.000 __init__.py:342(get_col)
   502793    2.051    0.000    7.118    0.000 expressions.py:657(as_sql)
1234731/315149    2.019    0.000    2.019    0.000 ElementTree.py:471(iter)
   497544    1.948    0.000    5.160    0.000 base_properties.py:224(value_to_property)
   830550    1.928    0.000    2.699    0.000 base.py:95(__getitem__)

```

after:
```
         80671674 function calls (76821377 primitive calls) in 272.946 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function) 
    52071   46.792    0.001   50.036    0.001 {method 'execute' of 'psycopg2.extensions.cursor' objects}
8289613/8289353    8.061    0.000    9.498    0.000 {isinstance}
  3429180    4.270    0.000    4.276    0.000 {hasattr}
  4560357    3.893    0.000    3.893    0.000 {method 'append' of 'list' objects}
    12704    3.875    0.000    4.827    0.000 clean_owners.py:111(_add_unchecked_indices_to_be_checked)
302527/12834    3.795    0.000    8.268    0.001 ElementTree.py:899(_serialize_xml) 
    24833    3.500    0.000    3.500    0.000 {method 'remove' of 'list' objects}
   616722    3.051    0.000    6.723    0.000 base_properties.py:224(value_to_property)
   705327    2.996    0.000    7.611    0.000 base_properties.py:71(__get__)
580174/527939    2.896    0.000   14.545    0.000 compiler.py:348(compile)
1941856/1913138    2.805    0.000    2.907    0.000 {getattr}
   732666    2.780    0.000    8.361    0.000 expressions.py:667(get_db_converters)
   772700    2.698    0.000    3.318    0.000 base_properties.py:20(__init__)
    54315    2.683    0.000    5.621    0.000 base.py:457(__init__)
  1030673    2.681    0.000    5.132    0.000 compiler.py:331(quote_name_unless_alias)
132428/41372    2.655    0.000   30.835    0.001 containers.py:111(__init__)
  1214072    2.583    0.000    2.585    0.000 {built-in method __new__ of type object at 0x919120}
    25699    2.346    0.000   21.676    0.001 compiler.py:165(get_select)
   150634    2.284    0.000    9.158    0.000 inspect.py:895(getcallargs)
      735    2.259    0.003    2.259    0.003 {built-in method read}
2487687/2097715    2.249    0.000   71.223    0.000 {len}
    12834    2.234    0.000    7.842    0.001 ElementTree.py:831(_namespaces)
    51327    2.099    0.000   11.115    0.000 compiler.py:764(get_converters)
    25691    2.092    0.000    6.678    0.000 compiler.py:475(get_default_columns)
    51141    2.086    0.000   50.897    0.001 query.py:1233(__iter__)
   502481    2.063    0.000    7.166    0.000 expressions.py:657(as_sql)
1234511/315109    2.013    0.000    2.013    0.000 ElementTree.py:471(iter)
   758974    1.951    0.000    2.463    0.000 __init__.py:342(get_col)
  1385477    1.884    0.000    1.884    0.000 {method 'items' of 'dict' objects}
    25629    1.829    0.000    8.404    0.000 query.py:1222(resolve_model_init_order)
    77433    1.741    0.000    3.705    0.000 query.py:258(clone)
   512840    1.731    0.000    3.466    0.000 query.py:1315(model_fields)
   293870    1.636    0.000    2.592    0.000 base_properties.py:241(value_to_python)
    25699    1.596    0.000   30.842    0.001 compiler.py:358(as_sql)
   732666    1.546    0.000    2.282    0.000 __init__.py:460(__eq__)
   840743    1.532    0.000    2.693    0.000 __init__.py:657(get_db_converters)
   696980    1.520    0.000    2.153    0.000 base.py:95(__getitem__)
281447/1915    1.408    0.000    3.719    0.002 copy.py:145(deepcopy)

```

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/phone/models.py [hunk 1]
```diff
@@ -700,10 +700,11 @@ def get_cases_that_directly_depend_on_case(self, case_id):
...
     def delete_index(self, from_case_id, index_name):
-        prior_ids = self.indices.pop(from_case_id, {})
-        prior_ids.pop(index_name, None)
-        if prior_ids:
-            self.indices[from_case_id] = prior_ids
+        case_indices = self.indices.get(from_case_id, {})
+        if len(case_indices) > 1:
+            case_indices.pop(index_name, {})
+        elif len(case_indices) == 1:
+            self.indices.pop(from_case_id)
```
Comment (proteusvacuum) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: This was an attempt to fix this:
```    
89240    0.617    0.000   31.671    0.000 models.py:708(indices)
```
but I don't think it worked, as that profile line didn't decrease when I made this change. 


the cumtime of something in here is 31.671s, which is nuts. It also looks like a large proportion of the time is taken up with `JsonObject` stuff, which I think is going to be hard to optimize around (other than reducing dict lookups).

Comment (proteusvacuum) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: aaand now that I write this comment, I notice that this is actually a reference to a different `models.py:708`

https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/models.py#L708-L717

**Review by calellowitz — COMMENTED** [review]
test failure looks real but otherwise looks good to me

File: corehq/ex-submodules/casexml/apps/phone/models.py [hunk 2]
```diff
@@ -623,46 +623,49 @@ class IndexTree(DocumentSchema):
     # and the values are the referenced case IDs
     indices = SchemaDictProperty()
 
+    _reverse_indices = None
+
+    @property
+    def reverse_indices(self):
+        if self._reverse_indices is None:
+            self._reverse_indices = _reverse_index_map(self.indices)
+        return self._reverse_indices
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: Why not use `@memoized`?

File: corehq/ex-submodules/casexml/apps/phone/models.py [hunk 3]
```diff
@@ -758,6 +759,16 @@ def purged_cases(self):
...
 
+    _frozen_closed_cases = None
+
+    @property
+    def frozen_closed_cases(self):
+        """Used so that we can memoize a function that requires this
+        """
+        if self._frozen_closed_cases is None:
+            self._frozen_closed_cases = frozenset(self.closed_cases)
+        return self._frozen_closed_cases
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: again, why not use `@memoized`?

**Review by snopoke — APPROVED** [review]

**Comment by proteusvacuum** [comment]
✅

**Pull request merged**